### PR TITLE
gettext: update to 0.21 with runtime and tools split

### PR DESCRIPTION
**Pull request by fredowski**
#### Description

- Update to upstream version 0.21
- Added two subports gettext-runtime and gettext-libtextstyle
- Removed the gettext-bootstrap subport
- Removed the myprefix construction

Notes:
- The gettext port can now install just the libintl library via the gettext-runtime subport.
- The gettext tools like msgfmt are installed via the gettext port.
- The tools have a dependency libtextstyle which is also added via a subport.
- I removed the bootstrap subport and the myprefix parts.

See: [Ticket 58526 - gettext: update to 0.21](https://trac.macports.org/ticket/58526)
See: [PR 7074 - python27-bootstrap: Stop using gettext-bootstrap](https://github.com/macports/macports-ports/pull/7074)
See: [Bug 58537 - MacOS: Default Language settings are not honored (gettext problem)](https://savannah.gnu.org/bugs/?58537)

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 10.13.6 17G12034
Xcode 10.1 10B61
###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: devel/gettext/Portfile
Comment (ryandesign) [inline devel/gettext/Portfile:9]: A port's epoch can never be decreased so `epoch 2` must stay, however ideally it should be in the `if {${subport} eq ${name}}` section so that it does not apply to the new subports being added here.

File: devel/gettext/Portfile
Comment (ryandesign) [inline devel/gettext/Portfile:27]: This comment should be removed.

File: devel/gettext/Portfile
Comment (ryandesign) [inline devel/gettext/Portfile:112]: Surely there must be a way to tell the build system not to build these again.

File: devel/gettext/Portfile
Comment (ryandesign) [inline devel/gettext/Portfile:11]: There should be a `revision 0` line in each subport and also in the `if {${subport} eq ${name}}` section.

File: devel/gettext/Portfile
Comment (fredowski) [reply devel/gettext/Portfile:9]: I added the epoch to the `if {${subport} eq ${name}}` section.

File: devel/gettext/Portfile
Comment (fredowski) [reply devel/gettext/Portfile]: I removed the comment.

File: devel/gettext/Portfile
Comment (fredowski) [reply devel/gettext/Portfile:114]: The worksrcdir is the gettext-tools subdirectory. So this should not build everything again. I had a look at the fink friends and it seems that the tools required some parts from runtime also. But this is how gettext is supposed to be build.

File: devel/gettext/Portfile
Comment (fredowski) [reply devel/gettext/Portfile:11]: I added a revision 0 in each subport and in the `if {${subport} eq ${name}}` section.

**Comment by fredowski** [comment]
Hi Ryan,

thanks for the ultrafast feedback! I added the parts (revision, epoch, comment) that you suggest. Regarding the rebuild - I build in the gettext-tools subdirectory so the libtextstyle part is not rebuild but is taken from the installation. Regarding the gettext-runtime this seems not to be so easy because some parts of gettext-runtime are touched. But this is what upstream suggests how to build the gettext-tools part. From macports perspective this is split in runtime and libtextstyle as dependency for gettext which is in fact gettext-tools.

**Comment by ryandesign** [comment]
I'll add commits to fix some issues I'm seeing. You can add commits too. We can squash it back down to 1 commit when we merge it to master.

**Comment by fredowski** [comment]
> I'll add commits to fix some issues I'm seeing. You can add commits too. We can squash it back down to 1 commit when we merge it to master.

O.k. I will not do any squashes and forced updates from now on.

Before this can be merged, you need to merge your removals of gettext-bootstrap
#7074 and #7075. I did not find any other places where gettext-bootstrap is used.

**Comment by fredowski** [comment]
Works for me with your commits.

**Comment by ryandesign** [comment]
Made a bunch of additional changes today, including moving libasprintf to its own subport. It sounds like a library that would be useful on its own and it doesn't sound very related to gettext. I thought about making it a dependency of gettext-runtime but decided not to when I couldn't find anywhere in gettext that it's used.

I spent a long time trying to get gettext-tools to stop also building and installing libintl again, but finally had to admit defeat and report it to the developers instead.

I updated the licenses of each subport. I think they're almost accurate now. The license of the gettext-runtime port is now `{LGPL-2.1+ GPL-3+}` as it was for the monolithic gettext before. That means the user may choose which of those two licenses they want to use. That wasn't accurate before and isn't accurate now. The software is under both licenses: libraries are under LGPL and programs are under GPL. But if we specified that (by writing `LGPL-2.1+ GPL-3+` without the curly braces) then MacPorts would interpret it as the most restrictive of those licenses and that would have caused a distributability problem. I think Josh probably opted to fudge the license here since the distributability checks are meant for libraries and the libintl library that everyone uses is LGPL. But not all of the libraries that were or are installed by gettext are LGPL.

Now that I've fixed the licenses of each port, the main gettext port that everybody depends on is GPL-3+. I think that's accurate but it will make some ports that depend on gettext nondistributable, for example:

```
"slrn" is not distributable because its license "GPL-2" conflicts with license "GPL-3+" of dependency "gettext"
```

This restriction would be accurate if slrn for example were using any of the libraries that remain in the gettext port: libgettextlib, libgettextpo, libgettextsrc. I assume nobody or at least very few ports do this. Almost all ports will only link with the LGPL libintl.

Each port that depends on gettext would need to be updated. The files installed by the port would have to be analyzed to determine if they are linking with the GPL libraries. Assuming they are not, the port could regain distributability by specifying `license_noconflict gettext`. Additionally ports that depend on gettext should change their dependencies. So a port that previously used:

```
depends_lib         port:gettext
```

would change it to:

```
depends_build       port:gettext
license_noconflict  gettext
depends_lib         port:gettext-runtime
```

Making this change in the almost 500 ports that use gettext isn't going to be fun. The ones where it affects distributability would be the ones to prioritize. That's at least the ports that use gettext and use license GPL-2, but maybe others as well.

We might be able to avoid having to require all ports to be audited and use `license_noconflict gettext` by adding another subport. We could split the GPL libraries into a new gettext-tools-libs subport. Then the gettext port would depend on gettext-tools-libs. But then the gettext port itself wouldn't install any libs anymore and could be marked `installs_libs no`. That should stop the license check from looking at the license of gettext or its dependencies.

**Comment by fredowski** [comment]
Wouldn't it be possible to make no license check for depends_build dependencies because that is by definition not linked to the runtime?

**Comment by jmroot** [comment]
> Wouldn't it be possible to make no license check for depends_build dependencies because that is by definition not linked to the runtime?

That is not the case. Things like headers and static libraries are only needed at build time but do produce a derivative work.

**Comment by ryandesign** [comment]
@jmroot do you agree with my assessment of the license issues? Do you see any other way around it?

File: devel/gettext/Portfile
Comment (jmroot) [inline devel/gettext/Portfile:24]: This option has done nothing since gcc 3.3 and TBH should have been removed years ago.

**Comment by jmroot** [comment]
> @jmroot do you agree with my assessment of the license issues? Do you see any other way around it?

I haven't looked at the licenses. I'll try to do so soon.

File: devel/gettext/Portfile
Comment (ryandesign) [reply devel/gettext/Portfile:24]: I was looking suspiciously at that line and couldn't remember if it was still needed on older systems. I've refreshed my memory now and agree with you, it should go. I should just take some time to remove it from all ports.

**Comment by jmroot** [comment]
@ryandesign: I looked now and your assessment seems correct.

> We might be able to avoid having to require all ports to be audited and use `license_noconflict gettext` by adding another subport. We could split the GPL libraries into a new gettext-tools-libs subport. Then the gettext port would depend on gettext-tools-libs. But then the gettext port itself wouldn't install any libs anymore and could be marked `installs_libs no`. That should stop the license check from looking at the license of gettext or its dependencies.

Sounds reasonable.

**Comment by cooljeanius** [comment]
so, newest version of gettext is now 0.21, not 0.20.2:

```
$ port -v livecheck gettext
gettext seems to have been updated (port version: 0.19.8.1, new version: 0.21)
```

**Comment by ryandesign** [comment]
Yes, it is. This PR should be updated to 0.21.

**Comment by fredowski** [comment]
I rebased on master cause 0.19 was updated in the meantime.

**Comment by fredowski** [comment]
I have made your proposed split in a separate gettext-tools-libs subport and did the update to 0.21.

**Comment by tessus** [comment]
qq is this dead?

I really need a newer version than 0.19.8.1. If this won't be merged soon, I'll have to compile it myself. This is not a problem, but I'd rather use macports.

**Comment by ryandesign** [comment]
@tessus It's not "dead" I just haven't tested the latest version of the PR and we haven't done the prerequisite work of removing the need for the removed gettext-bootstrap subport.

What aspect of newer gettext do you require?

**Comment by tessus** [comment]
In Joplin we use gettext for translations. However, JS templates aren't handle properly in gettext <0.21, and when it fails it skips the rest of the file, which means the translations get lost.

**Comment by mascguy** [comment]
Yes, I'd also love to see this upgrade for Mame: While the present version of gettext works, it also outputs plenty of bogus warnings due to lack of support for C++14. That was fixed as of v0.20:

[bug #50117: bogus "unterminated character constant" warning with C++14 single-quotation-mark as digit separators](https://savannah.gnu.org/bugs/?50117)

**Comment by watersb** [comment]
I am getting some test failures when running `port test gettext subport=gettext-tools-libs`.

One test was a hard fail in linked library, which tested for 64-bit ppc and x86 but not aarch64.

```
--- gettext-tools/gnulib-tests/vma-iter.c  2021-01-24 19:49:37.000000000 -0700
+++ gettext-tools/gnulib-tests/vma-iter.c  2021-01-24 19:44:50.000000000 -0700
@@ -1333,7 +1333,7 @@
          In 64-bit processes, we could use vm_region_64 or mach_vm_region.
          I choose vm_region_64 because it uses the same types as vm_region,
          resulting in less conditional code.  */
-# if defined __ppc64__ || defined __x86_64__
+# if defined __ppc64__ || defined __x86_64__ || defined __aarch64__
       struct vm_region_basic_info_64 info;
       mach_msg_type_number_t info_count = VM_REGION_BASIC_INFO_COUNT_64;
```

With that patched, all tests run, but I have 5 failures. My autoconf skills are rusty, so I'm taking a while to understand the build logs. So I don't yet know if these failures are relevant to the macOS platform.

```
:info:test PASS: test-fnmatch-h
:info:test PASS: test-fnmatch
:info:test ../../build-aux/test-driver: line 107: 79901 Abort trap: 6           "$@" > $log_file 2>&1
:info:test FAIL: test-fopen-gnu
:info:test PASS: test-fopen
:info:test PASS: test-fputc
:info:test PASS: test-fread
:info:test PASS: test-fstat
:info:test PASS: test-fstrcmp
:info:test FAIL: test-ftell.sh
:info:test ../../build-aux/test-driver: line 107: 80036 Abort trap: 6           "$@" > $log_file 2>&1
:info:test FAIL: test-ftell2.sh
:info:test PASS: test-ftell3
:info:test FAIL: test-ftello.sh
:info:test ../../build-aux/test-driver: line 107: 80094 Abort trap: 6           "$@" > $log_file 2>&1
:info:test FAIL: test-ftello2.sh
:info:test PASS: test-ftello3
:info:test PASS: test-ftello4.sh
:info:test PASS: test-ftruncate.sh
```

**Comment by cooljeanius** [comment]
> With that patched, all tests run, but I have 5 failures. My autoconf skills are rusty, so I'm taking a while to understand the build logs. So I don't yet know if these failures are relevant to the macOS platform.
> 
> ```
> :info:test PASS: test-fnmatch-h
> :info:test PASS: test-fnmatch
> :info:test ../../build-aux/test-driver: line 107: 79901 Abort trap: 6           "$@" > $log_file 2>&1
> :info:test FAIL: test-fopen-gnu
> :info:test PASS: test-fopen
> :info:test PASS: test-fputc
> :info:test PASS: test-fread
> :info:test PASS: test-fstat
> :info:test PASS: test-fstrcmp
> :info:test FAIL: test-ftell.sh
> :info:test ../../build-aux/test-driver: line 107: 80036 Abort trap: 6           "$@" > $log_file 2>&1
> :info:test FAIL: test-ftell2.sh
> :info:test PASS: test-ftell3
> :info:test FAIL: test-ftello.sh
> :info:test ../../build-aux/test-driver: line 107: 80094 Abort trap: 6           "$@" > $log_file 2>&1
> :info:test FAIL: test-ftello2.sh
> :info:test PASS: test-ftello3
> :info:test PASS: test-ftello4.sh
> :info:test PASS: test-ftruncate.sh
> ```

Pretty sure all those tests come from gnulib and I think I've seen them fail upstream, too, so I don't think they're really something MacPorts needs to fix. Feel free to report them to <bug-gnulib@gnu.org> though, if they haven't already been.

**Comment by mascguy** [comment]
Just checking in on this, as I'd love to see `gettext` updated to a current version. Are we potentially ready to push this...?

**Comment by mascguy** [comment]
Ryan, can one of us pick up the remaining work on this? Is it mainly testing/validation, or...?

**Comment by mascguy** [comment]
In the interest of making progress, I'm going to work on this.

Should I start with Boyd's work?
https://github.com/watersb/macports-ports/commit/7c364e374c838c406f8e20413c4799a1104bcc7c

Or what's currently on Friedrich's branch?

**Comment by mascguy** [comment]
FYI, just rebased branch against master. All looks fine, but let me know if anyone sees any issues.

**Comment by mascguy** [comment]
In terms of validating this port, I've successfully built `gettext` - and all subports (including `libasprintf`) - on the following macOS releases:

- 10.6.8
- 10.8.5
- 10.12.6
- 10.13.6
- 10.14.6
- 10.15.7

Unfortunately I'm not able to run Big Sur on my old MacPro yet, but the Big Sur CI job did run to completion.

I'm still planning to build these on macOS 10.10.5, for another data point. But so far, everything looks great!

So... what else is needed, apart from testing dependent [PR 7074 - python27-bootstrap: Stop using gettext-bootstrap](https://github.com/macports/macports-ports/pull/7074)?

**Review by mascguy — APPROVED** [review]

**Comment by mascguy** [comment]
FYI, clang-3.4 was rebuilt on macOS 10.6, with both gettext 0.21 and the updated python27-bootstrap (per the other PR), as requested by Josh.

What else is needed, in terms of validation, etc?

**Comment by jmroot** [comment]
This PR and #7074 aren't related except in that they both involve gettext.

**Comment by mascguy** [comment]
> This PR and #7074 aren't related except in that they both involve gettext.

Well, I'm simply trying to jump in, and help move these along. The precise relations aren't the point.

**Comment by mascguy** [comment]
With that being said, the associated python27 PR needs to be merged first, as this PR removes gettext-bootstrap.

**Comment by mascguy** [comment]
Rebased against master

**Comment by mascguy** [comment]
Rebased against master

**Comment by reneeotten** [comment]
@ryandesign what do you want to do with this PR? It's been open for more than a year now, so we should probably either work towards merging this or just close it if that's not going to happen.

**Comment by mascguy** [comment]
Recent work has been done to help test/validate these changes, along with the Python27 PR this depends on.

Ultimately we need another person to test that other PR, as it's required for this one.

CC: @kencu

**Comment by mascguy** [comment]
Rebased against master

**Comment by mascguy** [comment]
Now that [PR 7074 - python27-bootstrap: Stop using gettext-bootstrap](https://github.com/macports/macports-ports/pull/7074) has been merged, I believe (?) this can be as well.

But in the interest of not breaking anything, can folks provide a final blessing? @ryandesign and @jmroot?

**Comment by mascguy** [comment]
These changes have been in-play for the past six months, in every one of my macOS installations - physical and virtual, from 10.6 through 10.15.

And while that certainly doesn't guarantee that we won't find a problem somewhere, overall I'm very comfortable with merging this.

Any final thoughts/objections?

**Comment by mascguy** [comment]
I'd like to hear any final thoughts before 10:00pm EST today.

If I don't hear anything, I will merge.

**Comment by mascguy** [comment]
Final call for feedback requested via the Dev mailing list:

https://lists.macports.org/pipermail/macports-dev/2021-November/043937.html

**Comment by cooljeanius** [comment]
> Final call for feedback requested via the Dev mailing list:
> 
> https://lists.macports.org/pipermail/macports-dev/2021-November/043937.html

Only thing I'd look out for would be making sure that we get all the revbumps required for ports that depend on gettext, since there are a lot of them, but that can be done separately.

**Comment by mascguy** [comment]
Here we go. Finally merging!

**Comment by mascguy** [comment]
And a big Thank You go to @fredowski and @ryandesign for all of the work on this!

**Comment by cjones051073** [comment]
The commits here really should have been squashed when this PR was merged...

**Comment by mascguy** [comment]
> The commits here really should have been squashed when this PR was merged...

Well, I respectfully disagree.

Yes, the comments could have been reworded, per our commit guidelines. And the failure to do so is on me. (In retrospect, it wouldn't have taken much time, via a quick interactive rebase.)

But the history could be very helpful in the future, to understand precisely what was done, and why. And all of that detail would have been lost, had I chosen to squash all of these. So I stand by the decision, for better or worse.

**Comment by mascguy** [comment]
At the end of the day though, assertive folks like you and I are going to occasionally make decisions that ruffle some feathers. Despite trying to follow all procedures and guidelines, whilst still making progress. Hey, it happens!

But merging this long-overdue update is going to make a lot of folks very happy. In addition to fixing some key things, it also supports work that has been otherwise held back until now.

So let's celebrate the update - and indeed, this is cause for celebration - and move on!

**Comment by cjones051073** [comment]
> > The commits here really should have been squashed when this PR was merged...
> 
> Well, I respectfully disagree.
> 
> Yes, the comments could have been reworded, per our commit guidelines. And the failure to do so is on me. (In retrospect, it wouldn't have taken much time, via a quick interactive rebase.)
> 
> But the history could be very helpful in the future, to understand precisely what was done, and why. And all of that detail would have been lost, had I chosen to squash on commit. So I stand by the decision, for better or worse.

You really think a commit like this one

https://github.com/macports/macports-ports/commit/ff70b7f3d4b12977827833e650b2f2773e7ab0ba

deserves being retained forever in our master history ;)

Yes, probably squashing all the commits into a single one would have been too much, but I think at most there are only about 5-10 or so 'real' commits in this PR that deserved being kept in their own right. The majority where just 'this fixes some issue in some earlier commit' and as such should have been squashed out of existence.

Don't get me wrong - I do appreciate your pushing forward with PRs like this, and yes it is good to get it merged eventually. I just ask that the commit history in master is maintained in some sort of good order, as otherwise it will loose its usefulness, if we end up with it polluted with commits like the above.

**Comment by mascguy** [comment]
> You really think a commit like this one
> 
> [ff70b7f](https://github.com/macports/macports-ports/commit/ff70b7f3d4b12977827833e650b2f2773e7ab0ba)
> 
> deserves being retained forever in our master history ;)
> 
> Yes, probably squashing all the commits into a single one would have been too much, but I think at most there are only about 5-10 or so 'real' commits in this PR that deserved being kept in their own right. The majority where just 'this fixes some issue in some earlier commit' and as such should have been squashed out of existence.
> 
> Don't get me wrong - I do appreciate your pushing forward with PRs like this, and yes it is good to get it merged eventually. I just ask that the commit history in master is maintained in some sort of good order, as otherwise it will loose its usefulness, if we end up with it polluted with commits like the above.

Agreed Chris, non-compliant commit comments aren't ideal. And will certainly clean those up a bit before merge, if necessary, with future PRs like this.

But on a lighter note, yes, that particular comment was humorous enough to keep. And I'm sure Dan Quail would approve of the (mis)spelling!  ;-)

**Comment by fredowski** [comment]
Thanks for all your work!

**Comment by ryandesign** [comment]
> The commits here really should have been squashed when this PR was merged...

Yeah. I made individual commits with detailed commit messages to show Friedrich what I was doing, as it would have been impossible to tell had I just done everything at once. It was intended to help him understand what was wrong with the PR originally and how to fix each issue. I was also expecting others to push commits to this PR, so just squashing everything into one during the PR's development would have made that difficult. I did of course expect everything to be squashed into a single commit at the end with a proper commit message, as I said [above](https://github.com/macports/macports-ports/pull/7399#issuecomment-642483343).

But no matter. Thank you Chris for getting it merged finally.

**Pull request merged**
